src/document-fetcher.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {FetchInitDef, XhrBase, assertSuccess, setupInit} from './xhr-base';
+import {isFormDataWrapper} from './form-data-wrapper';
+import {user} from './log';
+
+export class DocumentFetcher extends XhrBase {
+  /**
+   * Creates an instance of DocumentFetcher.
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super(win);
+
+    /** @private {boolean} */
+    this.viewerResponded_ = false;
+
+    /** @private {!XMLHttpRequest} */
+    this.xhr_ = new XMLHttpRequest();
+  }
+
+  /**
+   * @override
+   */
+  fetchFromNetwork_(input, init) {
```

Comment:
I think trying to follow a rigid class hierarchy has really hurt this module. From my understanding, we need 4 things:

1. Setup input and init
2. Check for viewer intercept
3. Perform XHR
4. Verify CORS headers

I think trying to fit that in into `fetchFromNetwork_` and `fetchDocument` makes the logic really weird, and return the appropriate type so `fetchAmpCors_` can work, etc, just makes the everything really convoluted. Can we try separating out **1**, **2** and **4** into pure helper functions, and create a `fetchDocument` helper function (not a class) that will put them together correctly?

Reply:
> I would definitely be willing to move to helper functions instead, but post the split of two classes. Otherwise it looked like a gigantic refactor,

I think this PR is the giant refactor. 😉

> This would help with a better review and smaller PRs to keep everything a bit more trackable.

I think the helper functions first would actually solve this. Instead of introducing a common base class for these to inherit, we submit several PRs to split our XHR class into helper functions. Keep everything else the same, just whittle away what's implemented directly on the class. We should get to the point we can implement `fetchDocument` without any class hierarchy.

> but ended up up creating functions with a huge number of arguments, we need to pass the viewer, the ampdoc

Right now, we're creating unneeded wrappers just to satisfy the return types of the class methods. It leads to weird interactions like:

- https://github.com/ampproject/amphtml/pull/16726#discussion_r202476210
- https://github.com/ampproject/amphtml/pull/16726#discussion_r202480772
- https://github.com/ampproject/amphtml/pull/16726#discussion_r202481521
- https://github.com/ampproject/amphtml/pull/16726#discussion_r202483296